pytools/tag.py

```diff
@@ -130,4 +135,148 @@ class UniqueTag(Tag):
 
 # }}}
 
+
+# {{{ parse
+
+def build_tag_grammar(shortcuts):
+    tag_grammar_tmplt = """
+    tag: tag_class "(" params ")" -> map_tag
+    % if shortcuts:
+       | SHORTCUT                 -> map_shortcut
+    % endif
+
+    params:                       -> map_empty_args_params
+          | args                  -> map_args_only_params
+          | kwargs                -> map_kwargs_only_params
+          | args "," kwargs       -> map_args_kwargs_params
+
+    ?kwargs: kwarg
+           | kwargs "," kwarg     -> map_kwargs
+
+    args: arg                     -> map_singleton_args
+        | args "," arg            -> map_args
+
+    kwarg: name "=" arg           -> map_kwarg
+
+    ?arg: tag
+        | INT                     -> map_int
+        | ESCAPED_STRING          -> map_string
+
+    tag_class: module "." name    -> map_tag_class
+
+    module: name                  -> map_top_level_module
+          | module "." name       -> map_nested_module
+
+    name: CNAME                   -> map_name
+
+    % if shortcuts:
+    SHORTCUT: (${" | ".join('"' + name + '"' for name in shortcuts)})
```

Comment:
I think the grammar should be the same no matter what shortcuts are defined. If that makes the shortcuts slightly more cumbersome to type, so be it.

Reply:
Agreed, so the shortcuts are now prepended with a period (`"."`). That still allows them to be a convenient shortcut and also lets us having a well-defined grammar.